### PR TITLE
Add Regularization Modules for LSTM

### PR DESCRIPTION
**Pull request by Ashutosh-Adhikari**
Adding Regularization modules for LSTM.  Thank you! 

- [x]  Weight-dropping
- [x]  Embedding Dropout
- [x] Temporal Activation Regularization
- [x] Temporal Averaging 

CC: @daemon

### CODE REVIEW

File: lstm_regularization/__main__.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    else:
+        model = LSTMBaseline(config)
+        if args.cuda:
+            model.cuda()
+            print('Shift model to GPU')
+
+    parameter = filter(lambda p: p.requires_grad, model.parameters())
+    optimizer = torch.optim.Adam(parameter, lr=args.lr, weight_decay=args.weight_decay)
+
+    if args.dataset == 'SST-1':
```
Comment (daemon) [inline lstm_regularization/__main__.py]: Nit: could just do `table = {"SST-1": SST1, ...} ... table[...]`.

File: lstm_regularization/__main__.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        'model_outfile': args.save_path,   # actually a directory, using model_outfile to conform to Trainer naming convention
+        'logger': logger
+    }
+    trainer = TrainerFactory.get_trainer(args.dataset, model, None, train_iter, trainer_config, train_evaluator, test_evaluator, dev_evaluator)
+
+    if not args.trained_model:
+        trainer.train(args.epochs)
+    else:
+        if args.cuda:
+            model = torch.load(args.trained_model, map_location=lambda storage, location: storage.cuda(args.gpu))
```
Comment (daemon) [inline lstm_regularization/__main__.py]: Loading and saving the state dictionary is the recommended method:
https://pytorch.org/docs/stable/notes/serialization.html

File: lstm_regularization/__main__.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        evaluate_dataset('dev', Reuters, model, None, dev_iter, args.batch_size, args.gpu)
+        evaluate_dataset('test', Reuters, model, None, test_iter, args.batch_size, args.gpu)
+    elif args.dataset == 'AAPD':
+        evaluate_dataset('dev', AAPD, model, None, dev_iter, args.batch_size, args.gpu)
+        evaluate_dataset('test', AAPD, model, None, test_iter, args.batch_size, args.gpu)
+    else:
+        raise ValueError('Unrecognized dataset')
+
+    # Calculate dev and test metrics
+    if model.beta_ema>0:
```
Comment (daemon) [inline lstm_regularization/__main__.py]: PEP8

File: lstm_regularization/__main__.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+def evaluate_dataset(split_name, dataset_cls, model, embedding, loader, batch_size, device):
+    saved_model_evaluator = EvaluatorFactory.get_evaluator(dataset_cls, model, embedding, loader, batch_size, device)
+    scores, metric_names = saved_model_evaluator.get_scores()
+    logger.info('Evaluation metrics for {}'.format(split_name))
+    logger.info('\t'.join([' '] + metric_names))
+    logger.info('\t'.join([split_name] + list(map(str, scores))))
+
+
+if __name__ == '__main__':
```
Comment (daemon) [inline lstm_regularization/__main__.py]: Should move this entire code block to a `main()` function.

File: lstm_regularization/embed_regularize.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
```
Comment (daemon) [inline lstm_regularization/embed_regularize.py]: Please include the license from the original [AWD-LSTM repo](https://github.com/salesforce/awd-lstm-lm) for https://github.com/salesforce/awd-lstm-lm/blob/master/embed_regularize.py

File: lstm_regularization/locked_dropout.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+class LockedDropout(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, dropout=0.5):
+        if not self.training or not dropout:
+            return x
+        m = x.data.new(1, x.size(1), x.size(2)).bernoulli_(1 - dropout)
+        mask = Variable(m, requires_grad=False) / (1 - dropout)
```
Comment (daemon) [inline lstm_regularization/locked_dropout.py]: Variable deprecated in PyTorch 0.4

File: lstm_regularization/model.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import torch
+import torch.nn as nn
+
+import torch.nn.functional as F
+from copy import deepcopy
```
Comment (daemon) [inline lstm_regularization/model.py]: Wrong import order, see PEP8.
```
import python_modules

import third-party_modules

import local_project_modules
```

File: lstm_regularization/weight_drop.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import torch
```
Comment (daemon) [inline lstm_regularization/weight_drop.py]: License here as well.

File: lstm_regularization/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
+# lstm_baseline with Regularization
+
+Implementation of a standard LSTM using PyTorch and Torchtext for text classification baseline measurements with REgularization.
```
Comment (daemon) [inline lstm_regularization/README.md]: Nit: typo

File: lstm_regularization/__main__.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    # Set default configuration in : args.py
+    args = get_args()
+
+    # Set random seed for reproducibility
```
Comment (daemon) [inline lstm_regularization/__main__.py]: Should move this block into a common `set_seed` function.

File: lstm_regularization/model.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                self.fc1 = nn.Linear(config.hidden_dim, config.hidden_dim//2)   # Hidden Bottleneck Layer
+                self.fc2 = nn.Linear(config.hidden_dim//2, target_class)
+        else:
+            if self.is_bidirectional:
+                self.fc1 = nn.Linear(2 * config.hidden_dim, target_class)
+            else:
+                self.fc1 = nn.Linear(config.hidden_dim, target_class)
+        
+        if self.beta_ema>0:
+            print("Temporal Averaging here ;)")
```
Comment (daemon) [inline lstm_regularization/model.py]: Nit: remove ";)"

File: lstm_regularization/model.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            if self.TAR:
+                return self.fc2(x), rnn_outs.permute(1,0,2)
+            return self.fc2(x)
+        else:
+            if self.TAR:
+                return self.fc1(x), rnn_outs.permute(1,0,2)
+            return self.fc1(x)
+
+    def update_ema(self):
+        self.steps_ema+=1
```
Comment (daemon) [inline lstm_regularization/model.py]: PEP8

File: lstm_regularization/model.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            self.static_embed = nn.Embedding.from_pretrained(dataset.TEXT_FIELD.vocab.vectors, freeze=True)
+        elif config.mode == 'non-static':
+            self.non_static_embed = nn.Embedding.from_pretrained(dataset.TEXT_FIELD.vocab.vectors, freeze=False)
+        else:
+            print("Unsupported Mode")
+            exit()
+
+        self.lstm = nn.LSTM(config.words_dim, config.hidden_dim, dropout=config.dropout, num_layers=config.num_layers,
+                            bidirectional=self.is_bidirectional, batch_first=True)
+        ################
```
Comment (daemon) [inline lstm_regularization/model.py]: Don't need comments. Also PEP8 for `dropout = ...`

File: lstm_regularization/README.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+## Dataset
+
+We experiment the model on the following datasets.
+
+- Reuters dataset - ModApte splits
+
+## Settings
+
+Adadelta is used for training.
```
Comment (daemon) [inline lstm_regularization/README.md]: Adam, right?

File: lstm_regularization/model.py (hunk 11 above)
Comment (achyudhk) [reply lstm_regularization/model.py]: Also not sure why we need to print that to stdout

Comment (Ashutosh-Adhikari) [reply lstm_regularization/model.py]: Ah I forgot to remove that. Thanks!

File: lstm_regularization/model.py (hunk 13 above)
Comment (Ashutosh-Adhikari) [reply lstm_regularization/model.py]: Ported code from previous contributions for consistency. Nevertheless will change this. Thank you!
Will also remove the comments.

File: lstm_regularization/__main__.py (hunk 2 above)
Comment (Ashutosh-Adhikari) [reply lstm_regularization/__main__.py]: Yeah even I wanted to change this to `load_state_dict`. I'll try to add that for at least the regularization module. For other models as well we should change this at some later stage. Thank you!

Comment (achyudhk) [reply lstm_regularization/__main__.py]: I'll change my models to save just the state_dict too. Should make it easier to use the same code for check-pointing across models.

File: lstm_regularization/model.py (hunk 13 above)
Comment (daemon) [reply lstm_regularization/model.py]: Wrong line, see L41

**Comment by Ashutosh-Adhikari** [comment]
@daemon , Please check now. Thanks!

**Comment by daemon** [comment]
LGTM.
